File: server/models/discovery.go
package models

// PublicServerListItem is a discovery directory card. It exposes only public preview data —
// never members, channels, or private settings. OnlineCount is filled from the WS hub, not SQL.
type PublicServerListItem struct {
	ID               string  `json:"id"`
	Name             string  `json:"name"`
	IconURL          *string `json:"icon_url"`
	BannerURL        *string `json:"banner_url"`
	Description      *string `json:"description"`
	Category         *string `json:"category"`
	MemberCount      int     `json:"member_count"`
	OnlineCount      int     `json:"online_count"`
	Verified         bool    `json:"verified"`
	Featured         bool    `json:"featured"`
	ApprovalRequired bool    `json:"approval_required"`
	IsMember         bool    `json:"is_member"`
}

// PublicServerListParams filters and paginates the discovery list.
type PublicServerListParams struct {
	RequestingUserID string // for the is_member flag
	Category         string // empty = all categories
	Search           string // empty = no text filter
	FeaturedOnly     bool
	Limit            int
	Offset           int
}

// PublicServerListPage is a page of discovery results.
type PublicServerListPage struct {
	Items []PublicServerListItem `json:"items"`
	Total int                    `json:"total"`
}

File: client/src/api/discovery.ts
import { apiClient, uploadRequest, type UploadOptions } from "./client";
import type { Server } from "../types";

/** A public server card from the discovery directory (preview data only). */
export type PublicServerListItem = {
  id: string;
  name: string;
  icon_url: string | null;
  banner_url: string | null;
  description: string | null;
  category: string | null;
  member_count: number;
  online_count: number;
  verified: boolean;
  featured: boolean;
  approval_required: boolean;
  is_member: boolean;
};

export type PublicServerListPage = {
  items: PublicServerListItem[];
  total: number;
};

export type DiscoveryQuery = {
  q?: string;
  category?: string;
  featured?: boolean;
  excludeFeatured?: boolean;
  page?: number;
  limit?: number;
};

export async function listPublicServers(params: DiscoveryQuery) {
  const qs = new URLSearchParams();
  if (params.q) qs.set("q", params.q);
  if (params.category) qs.set("category", params.category);
  if (params.featured) qs.set("featured", "true");
  if (params.excludeFeatured) qs.set("exclude_featured", "true");
  if (params.page) qs.set("page", String(params.page));
  if (params.limit) qs.set("limit", String(params.limit));
  const query = qs.toString();
  return apiClient<PublicServerListPage>(`/discovery/servers${query ? `?${query}` : ""}`);
}

export async function getPublicServer(id: string) {
  return apiClient<PublicServerListItem>(`/discovery/servers/${id}`);
}

/** pending=true → an approval request was created (no membership yet). */
export async function joinPublicServer(id: string) {
  return apiClient<{ pending: boolean; server?: Server }>(`/discovery/servers/${id}/join`, {
    method: "POST",
  });
}

/** Report a public server for discovery moderation. Uses multipart when evidence files are given. */
export async function reportServer(
  id: string,
  reason: string,
  description: string,
  files?: File[],
  upload?: UploadOptions
) {
  if (files && files.length > 0) {
    const formData = new FormData();
    formData.append("reason", reason);
    formData.append("description", description);
    for (const file of files) {
      formData.append("files", file);
    }
    return uploadRequest<{ message: string }>(`/discovery/servers/${id}/report`, formData, upload);
  }
  return apiClient<{ message: string }>(`/discovery/servers/${id}/report`, {
    method: "POST",
    body: { reason, description },
  });
}
